File: Sources/ExcalidrawModel/History.swift
import Foundation

/// The change to a single element between two scene states. A `nil` side means
/// the element was absent (insertion when `before == nil`, removal when
/// `after == nil`).
public struct ElementChange: Equatable, Sendable {
    public var before: ExcalidrawElement?
    public var after: ExcalidrawElement?

    public init(before: ExcalidrawElement?, after: ExcalidrawElement?) {
        self.before = before
        self.after = after
    }
}

/// A reversible diff between two element lists, keyed by element id.
///
/// Phase 1 uses whole-element snapshots rather than upstream's property-level
/// partial deltas (`packages/element/src/delta.ts`). This is correct and simple
/// for local undo/redo; the finer-grained representation that collaboration
/// bandwidth needs is a later increment. AppState deltas are likewise deferred.
public struct SceneDelta: Equatable, Sendable {
    public var changes: [String: ElementChange]

    public init(changes: [String: ElementChange] = [:]) {
        self.changes = changes
    }

    public var isEmpty: Bool {
        changes.isEmpty
    }

    /// Diff `old` → `new`, recording only elements that actually changed.
    public static func between(_ old: [ExcalidrawElement], _ new: [ExcalidrawElement]) -> SceneDelta {
        let oldByID = Dictionary(old.map { ($0.id, $0) }, uniquingKeysWith: { _, last in last })
        let newByID = Dictionary(new.map { ($0.id, $0) }, uniquingKeysWith: { _, last in last })
        var changes: [String: ElementChange] = [:]
        for id in Set(oldByID.keys).union(newByID.keys) where oldByID[id] != newByID[id] {
            changes[id] = ElementChange(before: oldByID[id], after: newByID[id])
        }
        return SceneDelta(changes: changes)
    }

    /// The delta that undoes this one (`before` and `after` swapped).
    public func inverse() -> SceneDelta {
        SceneDelta(changes: changes.mapValues { ElementChange(before: $0.after, after: $0.before) })
    }
}

/// Undo/redo stacks of recorded deltas (`packages/excalidraw/history.ts`).
public struct History: Sendable {
    private var undoStack: [SceneDelta] = []
    private var redoStack: [SceneDelta] = []

    public init() {}

    public var canUndo: Bool {
        !undoStack.isEmpty
    }

    public var canRedo: Bool {
        !redoStack.isEmpty
    }

    /// Record a new change, clearing the redo stack (a fresh edit branches history).
    public mutating func record(_ delta: SceneDelta) {
        guard !delta.isEmpty else { return }
        undoStack.append(delta)
        redoStack.removeAll()
    }

    /// Pop the most recent change for undoing; the caller applies its inverse.
    public mutating func popUndo() -> SceneDelta? {
        guard let delta = undoStack.popLast() else { return nil }
        redoStack.append(delta)
        return delta
    }

    /// Pop the most recently undone change for redoing; the caller re-applies it.
    public mutating func popRedo() -> SceneDelta? {
        guard let delta = redoStack.popLast() else { return nil }
        undoStack.append(delta)
        return delta
    }
}

/// Ties a `Scene` to a `History`: capture edits, then undo/redo them. Mutate the
/// scene via `store.scene`, then call `commit()` to snapshot the change.
public struct Store: Sendable {
    public private(set) var scene: Scene
    private var history = History()
    private var snapshot: [ExcalidrawElement]

    public init(scene: Scene) {
        self.scene = scene
        snapshot = scene.elements
    }

    public var canUndo: Bool {
        history.canUndo
    }

    public var canRedo: Bool {
        history.canRedo
    }

    /// Mutate the scene in a closure and capture the change as one undo step.
    public mutating func transaction(_ body: (inout Scene) -> Void) {
        body(&scene)
        commit()
    }

    /// Mutate the working scene without recording history. Call `commit()` once
    /// the interaction settles to capture the net change as a single undo step.
    public mutating func modifyScene(_ body: (inout Scene) -> Void) {
        body(&scene)
    }

    /// Advance the undo baseline to the current scene *without* recording a step
    /// — used after applying a remote collaborative update so it doesn't fold
    /// into the local user's next undo. (parity: TS `Store.rebase`)
    public mutating func rebase() {
        snapshot = scene.elements
    }

    /// Capture any changes made to `scene` since the last commit as one undo step.
    public mutating func commit() {
        let delta = SceneDelta.between(snapshot, scene.elements)
        guard !delta.isEmpty else { return }
        history.record(delta)
        snapshot = scene.elements
    }

    @discardableResult
    public mutating func undo() -> Bool {
        guard let delta = history.popUndo() else { return false }
        scene.apply(delta.inverse())
        snapshot = scene.elements
        return true
    }

    @discardableResult
    public mutating func redo() -> Bool {
        guard let delta = history.popRedo() else { return false }
        scene.apply(delta)
        snapshot = scene.elements
        return true
    }
}

File: web/packages/excalidraw-svelte/src/model/history.ts
import type { ExcalidrawElement } from "./element.js";
import { semanticEqual } from "./json.js";
import type { Scene } from "./scene.js";

/**
 * The change to a single element between two scene states. A `null` side means
 * the element was absent (insertion when `before === null`, removal when
 * `after === null`). (parity: History.swift)
 */
export interface ElementChange {
  before: ExcalidrawElement | null;
  after: ExcalidrawElement | null;
}

function byId(list: ExcalidrawElement[]): Map<string, ExcalidrawElement> {
  const map = new Map<string, ExcalidrawElement>();
  for (const el of list) map.set(el.id, el);
  return map;
}

/** A reversible diff between two element lists, keyed by element id. */
export class SceneDelta {
  constructor(public changes: Map<string, ElementChange> = new Map()) {}

  get isEmpty(): boolean {
    return this.changes.size === 0;
  }

  /** Diff `old` → `next`, recording only elements that actually changed. */
  static between(oldList: ExcalidrawElement[], nextList: ExcalidrawElement[]): SceneDelta {
    const oldById = byId(oldList);
    const newById = byId(nextList);
    const changes = new Map<string, ElementChange>();
    for (const id of new Set([...oldById.keys(), ...newById.keys()])) {
      const before = oldById.get(id) ?? null;
      const after = newById.get(id) ?? null;
      if (!semanticEqual(before, after)) changes.set(id, { before, after });
    }
    return new SceneDelta(changes);
  }

  /** The delta that undoes this one (`before` and `after` swapped). */
  inverse(): SceneDelta {
    const inverted = new Map<string, ElementChange>();
    for (const [id, change] of this.changes) {
      inverted.set(id, { before: change.after, after: change.before });
    }
    return new SceneDelta(inverted);
  }
}

/** Undo/redo stacks of recorded deltas. */
export class History {
  private undoStack: SceneDelta[] = [];
  private redoStack: SceneDelta[] = [];

  get canUndo(): boolean {
    return this.undoStack.length > 0;
  }

  get canRedo(): boolean {
    return this.redoStack.length > 0;
  }

  /** Record a new change, clearing the redo stack (a fresh edit branches history). */
  record(delta: SceneDelta): void {
    if (delta.isEmpty) return;
    this.undoStack.push(delta);
    this.redoStack = [];
  }

  popUndo(): SceneDelta | null {
    const delta = this.undoStack.pop();
    if (delta === undefined) return null;
    this.redoStack.push(delta);
    return delta;
  }

  popRedo(): SceneDelta | null {
    const delta = this.redoStack.pop();
    if (delta === undefined) return null;
    this.undoStack.push(delta);
    return delta;
  }
}

/**
 * Ties a `Scene` to a `History`: capture edits, then undo/redo them. Mutate the
 * scene via `store.scene`, then `commit()` to snapshot the change.
 */
export class Store {
  private history = new History();
  private snapshot: ExcalidrawElement[];

  constructor(public scene: Scene) {
    this.snapshot = scene.elements;
  }

  get canUndo(): boolean {
    return this.history.canUndo;
  }

  get canRedo(): boolean {
    return this.history.canRedo;
  }

  /** Mutate the scene in a callback and capture the change as one undo step. */
  transaction(body: (scene: Scene) => void): void {
    body(this.scene);
    this.commit();
  }

  /** Mutate the working scene without recording history (live interaction). */
  modifyScene(body: (scene: Scene) => void): void {
    body(this.scene);
  }

  /**
   * Advance the undo baseline to the current scene *without* recording a step —
   * used after applying a remote collaborative update so it doesn't fold into
   * the local user's next undo.
   */
  rebase(): void {
    this.snapshot = this.scene.elements;
  }

  /** Capture changes since the last commit as one undo step. */
  commit(): void {
    const delta = SceneDelta.between(this.snapshot, this.scene.elements);
    if (delta.isEmpty) return;
    this.history.record(delta);
    this.snapshot = this.scene.elements;
  }

  undo(): boolean {
    const delta = this.history.popUndo();
    if (delta === null) return false;
    this.scene.apply(delta.inverse());
    this.snapshot = this.scene.elements;
    return true;
  }

  redo(): boolean {
    const delta = this.history.popRedo();
    if (delta === null) return false;
    this.scene.apply(delta);
    this.snapshot = this.scene.elements;
    return true;
  }
}
